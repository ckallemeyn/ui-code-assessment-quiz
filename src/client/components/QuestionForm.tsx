import React from 'react'
import { Container, Button, Form } from 'semantic-ui-react'
import TrueFalse from './TrueFalse'
import MultipleChoice from './MultipleChoice'
import ShortAnswer from './ShortAnswer'
import { unescapeStr } from '../../utils/questionUtils'

// const QUESTION_TYPES: any = {
//   boolean: <TrueFalse />,
//   multiple: <MultipleChoice />,
//   text: <ShortAnswer />
// }

// TODO: move answers array to a context for global consumption
// TODO: move QUESTION_TYPES enum out of component scope after context is created
const QuestionForm = ({ currentQuestion, answers, updateQuestion }: any) => {
  const { type } = currentQuestion;
  const question = unescapeStr(currentQuestion.question);

  const QUESTION_TYPES: any = {
    boolean: <TrueFalse />,
    multiple: <MultipleChoice answers={answers} />,
    text: <ShortAnswer />,
  };


  return (
    <>
      <Container textAlign="left">
        <Form onSubmit={updateQuestion}>
          <h3>{question}</h3>
          <div style={{ marginBottom: '10px' }}>
          {QUESTION_TYPES[type]}
          </div>
          <Button type="submit" primary>Next</Button>
        </Form>
      </Container>
    </>
  )
}

export default QuestionForm
