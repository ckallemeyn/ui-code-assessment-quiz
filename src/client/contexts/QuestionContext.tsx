import { createContext } from "react"

// maybe play around with creating an HOC for the entire app
const QuestionContext = createContext({});
QuestionContext.displayName = 'QuestionContext';

export default QuestionContext;
