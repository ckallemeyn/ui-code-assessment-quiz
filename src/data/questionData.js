export const questionData = [
  {
    category: "Entertainment: Video Games",
    type: "multiple",
    difficulty: "easy",
    question: "If a &quot;360 no-scope&quot; is one full rotation before shooting, how many rotations would a &quot;1080 no-scope&quot; be?",
    correct_answer: "3",
    incorrect_answers: [
      "4",
      "2",
      "5"
    ]
  },
  {
    category: "Animals",
    type: "text",
    difficulty: "medium",
    question: "What color\/colour is a polar bear&#039;s skin?",
    correct_answer: "Black"
  },
  {
    category: "Science & Nature",
    type: "boolean",
    difficulty: "easy",
    question: "Igneous rocks are formed by excessive heat and pressure.",
    correct_answer: "False",
    incorrect_answers: [
      "True"
    ]
  },
];
