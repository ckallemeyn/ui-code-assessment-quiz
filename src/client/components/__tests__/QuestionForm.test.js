import React from "react";
import { render } from "@testing-library/react";
import QuestionForm from "../QuestionForm";
import QuestionContext from '../../contexts/QuestionContext'
import { questionData } from '../../../data/questionData'

const renderWithContext = (value, props) => {
  return render(
    <QuestionContext.Provider value={value}>
      <QuestionForm {...props} />
    </QuestionContext.Provider>
  );
}

let contextValue;
let props;

beforeEach(() => {
  contextValue = {
    currentAnswer: '',
    handleChange: jest.fn(),
    state: {
      answers: ['4', '3', '2', '5']
    },
  };

  props = {
    updateQuestion: jest.fn(),
    state: {
      currentQuestion: questionData[0],
      currentAnswer: '',
      correctAnswers: 0,
      incorrectAnswers: 0,
      questionsAnswered: 0,
      finalScorePercentage: 0,
      isSummaryVisible: false,
    },
  };
});

describe('QuestionForm', () => {
  test('it should render', () => {
    const { container } = renderWithContext(contextValue, props);

    expect(container).toMatchSnapshot();
  });

  test('it should render a Multiple Choice question', () => {
    const { getByTestId } = renderWithContext(contextValue, props);
    const multipleChoiceQuestion = getByTestId('multiple');

    expect(multipleChoiceQuestion).toBeInTheDocument;
  });

  test('it should render a True/False question', () => {
    props.state.currentQuestion = questionData[2];

    const { getByTestId } = renderWithContext(contextValue, props);
    const trueFalseQuestion = getByTestId('boolean');

    expect(trueFalseQuestion).toBeInTheDocument;
  });

  test('it should render a Short Answer question', () => {
    props.state.currentQuestion = questionData[1];

    const { getByTestId } = renderWithContext(contextValue, props);
    const shortAnswerQuestion = getByTestId('text');

    expect(shortAnswerQuestion).toBeInTheDocument;
  });

  test('it should render the Summary page', () => {
    props.state.isSummaryVisible = true;
    
    const { getByTestId } = renderWithContext(contextValue, props);
    const summaryPage = getByTestId('summary');

    expect(summaryPage).toBeInTheDocument;
  });

});
