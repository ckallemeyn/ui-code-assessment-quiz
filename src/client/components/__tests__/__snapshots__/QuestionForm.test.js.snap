// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionForm it should render 1`] = `
<div>
  <div
    class="ui left aligned container"
  >
    <form
      class="ui form"
      data-testid="question-form"
    >
      <h3
        data-testid="question-text"
      >
        If a "360 no-scope" is one full rotation before shooting, how many rotations would a "1080 no-scope" be?
      </h3>
      <div
        class="question-type"
        data-testid="multiple"
        style="margin-bottom: 10px;"
      >
        <div
          class="field"
          data-testid="multiple-choice-answer"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              name="radioGroup"
              readonly=""
              tabindex="0"
              type="radio"
              value="4"
            />
            <label>
              4
            </label>
          </div>
        </div>
        <div
          class="field"
          data-testid="multiple-choice-answer"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              name="radioGroup"
              readonly=""
              tabindex="0"
              type="radio"
              value="3"
            />
            <label>
              3
            </label>
          </div>
        </div>
        <div
          class="field"
          data-testid="multiple-choice-answer"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              name="radioGroup"
              readonly=""
              tabindex="0"
              type="radio"
              value="2"
            />
            <label>
              2
            </label>
          </div>
        </div>
        <div
          class="field"
          data-testid="multiple-choice-answer"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              name="radioGroup"
              readonly=""
              tabindex="0"
              type="radio"
              value="5"
            />
            <label>
              5
            </label>
          </div>
        </div>
      </div>
      <button
        class="ui primary disabled button"
        data-testid="next-btn"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        Next
      </button>
    </form>
  </div>
</div>
`;
